src/config/source/yaml_file.cpp

```diff
@@ -0,0 +1,273 @@
+   // Push the last accumulated string to result
+   if (!current.empty()) {
+      result.push_back(current);
+   }
+
+   return result;
+}
+
+std::string joinCamelCase(const std::vector<std::string>& words) {
+   std::string camel_case_string;
+
+   for (size_t i = 0; i < words.size(); ++i) {
+      if (i == 0) {
+         // Add the first word as is (lowercase)
+         camel_case_string += words[i];
+      } else {
+         // Capitalize the first character of subsequent words and append them
+         std::string word = words[i];
+         if (!word.empty()) {
+            word[0] = static_cast<char>(std::toupper(word[0]));
+            camel_case_string += word;
+         }
+      }
+   }
+
+   return camel_case_string;
+}
+
+// NOLINTNEXTLINE(misc-no-recursion)
+void yamlToPaths(
+   const std::string& config_context,
+   const YAML::Node& node,
+   const ConsList<std::vector<std::string>>& parents,
+   std::unordered_map<ConfigKeyPath, YAML::Node>& paths
+) {
+   if (node.IsMap()) {
+      for (const auto& key_value : node) {
+         try {
+            const auto key = key_value.first.as<std::string>();
+            const auto parents2 = parents.cons(splitCamelCase(key));
+            const auto child_node = key_value.second;
+            yamlToPaths(config_context, child_node, parents2, paths);
+         } catch (YAML::BadConversion& bad_conversion) {
+            throw silo::config::ConfigException(fmt::format(
+               "invalid (non-literal) key in yaml config file: {}", bad_conversion.what()
+            ));
+         }
+      }
+   } else {
+      auto parents_vector = parents.toVecReverse();
+      auto path = ConfigKeyPath::tryFrom(parents_vector);
+      if (!path) {
+         // transform the ConsList parents into a debug string by reversing it,
+         // transforming the elements into camel-case and joining them with a '.'
+         auto debug_string_parents = ({
+            std::vector<std::string> result;
+            std::ranges::transform(parents_vector, std::back_inserter(result), joinCamelCase);
+            boost::join(result, ": ");
+         });
+         throw silo::config::ConfigException(
+            fmt::format("{}: found invalid key: {}", config_context, debug_string_parents)
+         );
+      }
+      if (isProperSingularValue(node)) {
+         paths.emplace(path.value(), node);
+      } else {
+         throw silo::config::ConfigException(fmt::format(
+            "{}: found non-usable leaf value at nesting {}",
+            config_context,
+            silo::config::YamlFile::configKeyPathToString(path.value())
+         ));
+      }
+   }
+}
+
+}  // namespace
+
+namespace silo::config {
+
+std::string YamlFile::configKeyPathToString(const ConfigKeyPath& config_key_path) {
+   std::vector<std::string> camel_case_strings;
+   for (const auto& list : config_key_path.getPath()) {
+      camel_case_strings.emplace_back(joinCamelCase(list));
+   }
+   return boost::join(camel_case_strings, ": ");
```

Comment:
I think the colon will confuse people even more, because that's not how you configure it either (since it's missing the newline and the indent after the colon).

The Loculus docs did it with the dot notation and it feels very natural: https://loculus.org/reference/helm-chart-config/